### PR TITLE
feat(gar): optimize prescribed observers, use lookup table and comput…

### PR DESCRIPTION
**Pull request by dtfiedler**
…e full array on read handlers

### CODE REVIEW

File: src/main.lua [hunk 1]
```diff
@@ -597,6 +597,12 @@ addEventingHandler(ActionMap.Transfer, utils.hasMatchingTag("Action", ActionMap.
 		msg.ioEvent:addField("RecipientNewBalance", recipientNewBalance)
 	end
 
+	-- if the sender is the protocol, then we need to update the circulating supply as tokens are now in circulation
+	if msg.From == ao.id then
+		LastKnownCirculatingSupply = LastKnownCirculatingSupply + quantity
+		addSupplyData(msg.ioEvent)
+	end
```
Comment (fedellen) [inline src/main.lua:604]: hmm, saw some test errors when looking into that `beforeEAch` mystery that might be related to this fix

File: src/main.lua [hunk 2]
```diff
@@ -1903,12 +1909,8 @@ end)
 
 addEventingHandler(ActionMap.Epoch, utils.hasMatchingTag("Action", ActionMap.Epoch), function(msg)
 	-- check if the epoch number is provided, if not get the epoch number from the timestamp
-	local providedEpochIndex = msg.Tags["Epoch-Index"]
-	local timestamp = msg.Timestamp
-
-	assert(providedEpochIndex or timestamp, "Epoch index or timestamp is required")
-
-	local epochIndex = providedEpochIndex or epochs.getEpochIndexForTimestamp(timestamp)
+	local epochIndex = msg.Tags["Epoch-Index"] and tonumber(msg.Tags["Epoch-Index"])
```
Comment (fedellen) [inline src/main.lua:1912]: is epoch-index already sanitized as number?

**Review by fedellen — COMMENTED** [review]
Looks good -- I think its ready but needs re-integration with latest develop

File: src/main.lua (hunk 2 above)
Comment (dtfiedler) [reply src/main.lua:1896]: https://github.com/ar-io/ar-io-network-process/blob/develop/src/utils.lua#L676

**Review by fedellen — DISMISSED** [review]

**Comment by dtfiedler** [comment]
just FYI - it'll take 14 epochs to prune out all the old state of previous epochs. As mentioned, this type is not used externally anywhere so no breaking change.

**Review by fedellen — APPROVED** [review]

**Pull request merged**